File: test/uk/co/bbc/dojo/DecoderTest.java
package uk.co.bbc.dojo;

import org.junit.Test;

import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertThat;

public class DecoderTest {

    private MediaFactory factory = new MediaFactory();

    private Decoder underTest = new Decoder();

    @Test
    public void shouldDecodeSimpleAudioByLowerCase() {
        Media media =  factory.create("simple","AUDIO","video");

        String decodedAudio = underTest.decodeAudio(media);

        assertThat(decodedAudio,is("audio"));
    }

    @Test
    public void shouldDecodeSimpleVideoByUpperCase() {
        Media media =  factory.create("simple","AUDIO","video");

        String decodedAudio = underTest.decodeVideo(media);

        assertThat(decodedAudio,is("VIDEO"));
    }

    @Test
    public void shouldDecodeAdvancedAudioByExpandingAndLowerCase() {
        Media media =  factory.create("advancedAudio","AUDIO","video");

        String decodedAudio = underTest.decodeAudio(media);

        assertThat(decodedAudio,is("aauuddiioo"));
    }

    @Test
    public void shouldDecodeVideoFromAdvancedAudioNormally() {
        Media media =  factory.create("advancedAudio","AUDIO","video");

        String decodedAudio = underTest.decodeVideo(media);

        assertThat(decodedAudio,is("VIDEO"));
    }


    @Test
    public void shouldDecodeVideoFromAdvancedBothByUpperCaseEveryOtherLetter() {
        Media media =  factory.create("advancedBoth","AUDIO","video");

        String decodedAudio = underTest.decodeVideo(media);

        assertThat(decodedAudio,is("ViDeO"));
    }

    @Test
    public void shouldDecodeAudioFromAdvancedBothByExpandAndLowerCase() {
        Media media =  factory.create("advancedBoth","AUDIO","video");

        String decodedAudio = underTest.decodeAudio(media);

        assertThat(decodedAudio,is("aauuddiioo"));
    }

    @Test
    public void shouldDecodeAdvancedVideoByUpperCaseEveryOtherLetter() {
        Media media =  factory.create("advancedVideo","AUDIO","video");

        String decodedAudio = underTest.decodeVideo(media);

        assertThat(decodedAudio,is("ViDeO"));
    }

    @Test
    public void shouldDecodeAudioFromAdvancedVideoNormally() {
        Media media =  factory.create("advancedVideo","AUDIO","video");

        String decodedAudio = underTest.decodeAudio(media);

        assertThat(decodedAudio,is("audio"));
    }
}

File: scala/src/test/scala/uk/co/bbc/dojo/DecoderTest.scala
package uk.co.bbc.dojo

import org.hamcrest.CoreMatchers._
import org.junit.Assert._
import org.junit.Test

class DecoderTest {
  private val factory: MediaFactory = new MediaFactory
  private val underTest: Decoder = new Decoder

  @Test
  def shouldDecodeSimpleAudioByLowerCase() = {
    val media: Media = factory.create("simple", "AUDIO", "video")
    val decodedAudio: String = underTest.decodeAudio(media)
    assertThat(decodedAudio, is("audio"))
  }

  @Test def shouldDecodeSimpleVideoByUpperCase() = {
    val media: Media = factory.create("simple", "AUDIO", "video")
    val decodedAudio: String = underTest.decodeVideo(media)
    assertThat(decodedAudio, is("VIDEO"))
  }

  @Test def shouldDecodeAdvancedAudioByExpandingAndLowerCase() = {
    val media: Media = factory.create("advancedAudio", "AUDIO", "video")
    val decodedAudio: String = underTest.decodeAudio(media)
    assertThat(decodedAudio, is("aauuddiioo"))
  }

  @Test def shouldDecodeVideoFromAdvancedAudioNormally() = {
    val media: Media = factory.create("advancedAudio", "AUDIO", "video")
    val decodedAudio: String = underTest.decodeVideo(media)
    assertThat(decodedAudio, is("VIDEO"))
  }

  @Test def shouldDecodeVideoFromAdvancedBothByUpperCaseEveryOtherLetter() = {
    val media: Media = factory.create("advancedBoth", "AUDIO", "video")
    val decodedAudio: String = underTest.decodeVideo(media)
    assertThat(decodedAudio, is("ViDeO"))
  }

  @Test def shouldDecodeAudioFromAdvancedBothByExpandAndLowerCase() = {
    val media: Media = factory.create("advancedBoth", "AUDIO", "video")
    val decodedAudio: String = underTest.decodeAudio(media)
    assertThat(decodedAudio, is("aauuddiioo"))
  }

  @Test def shouldDecodeAdvancedVideoByUpperCaseEveryOtherLetter() = {
    val media: Media = factory.create("advancedVideo", "AUDIO", "video")
    val decodedAudio: String = underTest.decodeVideo(media)
    assertThat(decodedAudio, is("ViDeO"))
  }

  @Test def shouldDecodeAudioFromAdvancedVideoNormally() = {
    val media: Media = factory.create("advancedVideo", "AUDIO", "video")
    val decodedAudio: String = underTest.decodeAudio(media)
    assertThat(decodedAudio, is("audio"))
  }
}